### PR TITLE
add Array.map_inplace

### PR DESCRIPTION
**Pull request by redianthus**
Hi,

This PR adds a `val map_inplace : ('a -> 'a) -> 'a array -> unit` function to the `Array` module.

This allow to replace code like :

```ocaml
let a = Array.init 10 (fun n -> 2 * n)
let () = Array.iteri (fun i x ->
    Array.set a i (succ x)
  ) a
```

by :

```ocaml
let a = Array.init 10 (fun n -> 2 * n)
let () = Array.map_inplace succ a
```

The function is in containers and in base, with the exact same name. In base it's implemented the same way, in containers it's using `iteri` instead of a loop.

I've written a simple test and added the function in the `ArrayLabels` module too.

The CI says:

> ./stdlib/array.mli:166.16: [doc-version] version should be formatted M.m (e.g. 5.0)

That's because I wrote `5.01`. But I'm not sure which one is the best as everywhere else it's e.g. `4.03` or `4.07` (in `array.mli` at least).

### CODE REVIEW

**Comment by dra27** [comment]
> That's because I wrote `5.01`. But I'm not sure which one is the best as everywhere else it's e.g. `4.03` or `4.07` (in `array.mli` at least).

CI knows best 🙂 (see #11683 and many instances of `@since 5.1` already merged)

**Comment by redianthus** [comment]
Fixed. ;)

**Review by gasche — APPROVED** [review]
The suggestion seems very reasonablee to me.

(Note: stdlib additions require two approvals.)

**Review by kayceesrk — APPROVED** [review]
This does look like a useful addition. The code looks right.

File: stdlib/arrayLabels.mli
Comment (yallop) [inline stdlib/arrayLabels.mli]:
```suggestion
(** [map_inplace ~f a] applies function [f] to all elements of [a],
```

File: stdlib/arrayLabels.mli
Comment (yallop) [inline stdlib/arrayLabels.mli]:
```suggestion
    and updates their values in place.
```

File: stdlib/array.mli
Comment (yallop) [inline stdlib/array.mli]:
```suggestion
    and updates their values in place.
```

File: stdlib/arrayLabels.mli
Comment (redianthus) [reply stdlib/arrayLabels.mli]: Fixed, thanks !

File: stdlib/arrayLabels.mli
Comment (redianthus) [reply stdlib/arrayLabels.mli]: Fixed.

File: stdlib/array.mli
Comment (redianthus) [reply stdlib/array.mli]: Fixed.

**Comment by alainfrisch** [comment]
This was fast :-)

FTR, two comments:

   - `Array.mapi_inplace` (looking nicer than `Array.map_inplacei`) might be useful as well.
   - What about avoiding the write if the result is physically equal to the original value? (To be confirmed on actual payload that this is indeed beneficial, and that the extra check does not degrade performance when it doesn't trigger the optim.)

**Comment by redianthus** [comment]
Thanks !

**Comment by redianthus** [comment]
@alainfrisch, I've opened #11837 with `mapi_inplace`.

About avoiding the write when the result is physically equal to the previous value, I don't have the time to properly benchmark it, so I'll let someone else look into it.

**Comment by nojb** [comment]
> * What about avoiding the write if the result is physically equal to the original value? (To be confirmed on actual payload that this is indeed beneficial, and that the extra check does not degrade performance when it doesn't trigger the optim.)

I did a quick benchmark with the following program:
```
let map_inplace a f =
  for i = 0 to Array.length a - 1 do
    Array.unsafe_set a i (f (Array.unsafe_get a i))
  done

let map_inplace' a f =
  for i = 0 to Array.length a - 1 do
    let x = Array.unsafe_get a i in
    let x' = f x in
    if x != x' then
      Array.unsafe_set a i (f (Array.unsafe_get a i))
  done

let test n init f =
  let a = Array.init n init in
  let t0 = Unix.gettimeofday () in
  for i = 1 to 1000 do
    map_inplace a f
  done;
  let t1 = Unix.gettimeofday () in
  for i = 1 to 1000 do
    map_inplace' a f
  done;
  let t2 = Unix.gettimeofday () in
  t1 -. t0, t2 -. t1

let test s init f =
  Printf.printf "\n%s\nn\tnoopt\topt\n" s;
  List.iter (fun n ->
      let t, t' = test n init f in
      Printf.printf "%d\t%.2fs\t%.2fs\n" n t t'
    ) [10; 100; 1000; 10000; 100000]

let () =
  test "succ(imm)" (fun i -> i) succ;
  test "id(imm)" (fun i -> i) Fun.id;
  test "succ+id(imm)" (fun i -> i) (fun n -> if n mod 2 == 0 then n else n + 1);
  test "succ(ptr)" (fun i -> i, "hello") (fun (i, s)  -> (i + 1, s));
  test "id(ptr)" (fun i -> i, "hello") Fun.id;
  test "succ+id(ptr)" (fun i -> i, "hello") (fun ((i, s) as x) -> if i mod 2 == 0 then (i + 1, s) else x)
```

The results are as follows (in my machine, in a single run). The column `n` is the size of the array being mapped, `noopt` menas the version of `map_inplace` in this PR, `opt` means the one with the "do not modify if physically equal" optimization. `succ` means that the function being mapped is `succ` (so the optimization does not trigger), `id` maps with the identity (so the optimization triggers for every element), `succ+id` triggers the optimization about half the time. Finally, `(imm)` means the array is made up of integers, and `(ptr)` that it is made up of heap-allocated values.

The results show a nice speedup when the optimization triggers, the speedup being most marked for heap-allocated values. The slowdown when the optimization does _not_ trigger is between 10% and 15%.

```
succ(imm)
n       noopt   opt
10      0.00s   0.00s
100     0.00s   0.00s
1000    0.00s   0.00s
10000   0.03s   0.04s
100000  0.34s   0.39s

id(imm)
n       noopt   opt
10      0.00s   0.00s
100     0.00s   0.00s
1000    0.00s   0.00s
10000   0.03s   0.01s
100000  0.27s   0.18s

succ+id(imm)
n       noopt   opt
10      0.00s   0.00s
100     0.00s   0.00s
1000    0.00s   0.00s
10000   0.04s   0.02s
100000  0.31s   0.17s

succ(ptr)
n       noopt   opt
10      0.00s   0.00s
100     0.00s   0.00s
1000    0.00s   0.01s
10000   0.07s   0.13s
100000  3.35s   3.72s

id(ptr)
n       noopt   opt
10      0.00s   0.00s
100     0.00s   0.00s
1000    0.00s   0.00s
10000   0.03s   0.01s
100000  0.55s   0.15s

succ+id(ptr)
n       noopt   opt
10      0.00s   0.00s
100     0.00s   0.00s
1000    0.00s   0.00s
10000   0.04s   0.02s
100000  0.44s   0.24s
```

**Pull request merged**
